/*
 * CMS Integration gRPC Service Definitions
 * 
 * This protobuf defines the gRPC contracts for SyncPortal to communicate
 * with the Council of Nine agents for closed-loop revenue optimization.
 * 
 * Services:
 * - CMSDataService: Push CMS data to other agents
 * - CircuitBreakerService: Control campaign pause/resume via SyncFlow
 * - AttributionService: OaaS settlement via SyncLedger
 * - CreativeAssetService: Product feed sync to SyncCreate
 */

syntax = "proto3";

package kiki.syncportal;

import "google/protobuf/timestamp.proto";

// ============================================================================
// Product Data Structures
// ============================================================================

message Product {
    string platform_product_id = 1;
    string sku = 2;
    string title = 3;
    string description = 4;
    double price = 5;
    int32 inventory_quantity = 6;
    bool in_stock = 7;
    string image_url = 8;
    repeated string additional_images = 9;
    string product_url = 10;
    string vendor = 11;
    string product_type = 12;
    bool is_active = 13;
    double cost_per_item = 14;
    google.protobuf.Timestamp updated_at = 15;
}

message Customer {
    string platform_customer_id = 1;
    string email = 2;
    string first_name = 3;
    string last_name = 4;
    string phone = 5;
    int32 total_orders = 6;
    double total_spent = 7;
    google.protobuf.Timestamp first_order_date = 8;
    google.protobuf.Timestamp last_order_date = 9;
    double predicted_ltv = 10;
    double churn_risk_score = 11;
}

message OrderItem {
    string platform_line_item_id = 1;
    string platform_product_id = 2;
    string sku = 3;
    string product_title = 4;
    int32 quantity = 5;
    double price = 6;
    double total_price = 7;
}

message Order {
    string platform_order_id = 1;
    string order_number = 2;
    double total_price = 3;
    double subtotal_price = 4;
    double total_tax = 5;
    double total_shipping = 6;
    double total_discounts = 7;
    string currency = 8;
    string financial_status = 9;
    string fulfillment_status = 10;
    google.protobuf.Timestamp order_date = 11;
    Customer customer = 12;
    repeated OrderItem items = 13;
}

// ============================================================================
// SyncValue™ Integration (LTV Predictions)
// ============================================================================

service LTVPredictionService {
    // Update customer LTV based on new order data
    rpc UpdateCustomerLTV(UpdateLTVRequest) returns (UpdateLTVResponse);
    
    // Batch update for historical baseline calculation
    rpc BatchUpdateBaseline(BatchUpdateBaselineRequest) returns (BatchUpdateBaselineResponse);
    
    // Get LTV prediction for attribution confidence
    rpc GetCustomerLTV(GetLTVRequest) returns (GetLTVResponse);
}

message UpdateLTVRequest {
    int32 store_id = 1;
    Customer customer = 2;
    Order latest_order = 3;
    repeated Order historical_orders = 4;
}

message UpdateLTVResponse {
    bool success = 1;
    double predicted_ltv = 2;
    double confidence = 3;
    double churn_risk_score = 4;
    string error = 5;
}

message BatchUpdateBaselineRequest {
    int32 store_id = 1;
    repeated Order historical_orders = 2;
    google.protobuf.Timestamp baseline_start_date = 3;
    google.protobuf.Timestamp baseline_end_date = 4;
}

message BatchUpdateBaselineResponse {
    bool success = 1;
    double baseline_revenue = 2;
    int32 baseline_order_count = 3;
    double avg_order_value = 4;
    string error = 5;
}

message GetLTVRequest {
    int32 store_id = 1;
    string customer_email = 2;
}

message GetLTVResponse {
    bool success = 1;
    double predicted_ltv = 2;
    double churn_risk_score = 3;
    string error = 4;
}

// ============================================================================
// SyncFlow™ Integration (Circuit Breaker)
// ============================================================================

service CircuitBreakerService {
    // Pause campaigns when product goes out of stock
    rpc PauseProductCampaigns(PauseCampaignsRequest) returns (PauseCampaignsResponse);
    
    // Resume campaigns when product is restocked
    rpc ResumeProductCampaigns(ResumeCampaignsRequest) returns (ResumeCampaignsResponse);
    
    // Adjust bids based on inventory levels
    rpc AdjustBidsByInventory(AdjustBidsRequest) returns (AdjustBidsResponse);
}

message PauseCampaignsRequest {
    int32 store_id = 1;
    int32 product_id = 2;
    string sku = 3;
    string reason = 4; // "out_of_stock", "low_inventory"
    string triggered_by = 5; // Webhook ID or manual
}

message PauseCampaignsResponse {
    bool success = 1;
    repeated string paused_campaign_ids = 2;
    int32 campaigns_affected = 3;
    string error = 4;
}

message ResumeCampaignsRequest {
    int32 store_id = 1;
    int32 product_id = 2;
    string sku = 3;
    int32 new_inventory_level = 4;
}

message ResumeCampaignsResponse {
    bool success = 1;
    repeated string resumed_campaign_ids = 2;
    int32 campaigns_affected = 3;
    string error = 4;
}

message AdjustBidsRequest {
    int32 store_id = 1;
    int32 product_id = 2;
    int32 inventory_level = 3;
    int32 low_stock_threshold = 4;
    double bid_adjustment_factor = 5; // 0.5 = reduce by 50%
}

message AdjustBidsResponse {
    bool success = 1;
    int32 campaigns_adjusted = 2;
    double new_bid_multiplier = 3;
    string error = 4;
}

// ============================================================================
// SyncLedger™ Integration (OaaS Attribution)
// ============================================================================

service AttributionService {
    // Record attributed order for OaaS billing
    rpc RecordAttributedOrder(RecordOrderRequest) returns (RecordOrderResponse);
    
    // Calculate incremental revenue for settlement
    rpc CalculateIncremental(CalculateIncrementalRequest) returns (CalculateIncrementalResponse);
    
    // Get attribution confidence for an order
    rpc GetAttributionConfidence(GetAttributionRequest) returns (GetAttributionResponse);
}

message RecordOrderRequest {
    int32 store_id = 1;
    Order order = 2;
    double attribution_confidence = 3;
    bool is_incremental = 4;
    string campaign_id = 5;
    repeated string touchpoint_ids = 6;
}

message RecordOrderResponse {
    bool success = 1;
    double incremental_revenue = 2;
    double success_fee = 3; // 20% of incremental
    string error = 4;
}

message CalculateIncrementalRequest {
    int32 store_id = 1;
    google.protobuf.Timestamp period_start = 2;
    google.protobuf.Timestamp period_end = 3;
}

message CalculateIncrementalResponse {
    bool success = 1;
    double baseline_revenue = 2;
    double current_period_revenue = 3;
    double incremental_revenue = 4;
    double growth_percentage = 5;
    int32 attributed_orders = 6;
    string error = 7;
}

message GetAttributionRequest {
    int32 store_id = 1;
    string customer_email = 2;
    google.protobuf.Timestamp order_date = 3;
}

message GetAttributionResponse {
    bool success = 1;
    double confidence = 2;
    bool attributed_to_kiki = 3;
    repeated string touchpoint_ids = 4;
    string campaign_id = 5;
    string error = 6;
}

// ============================================================================
// SyncCreate™ Integration (Creative Asset Sync)
// ============================================================================

service CreativeAssetService {
    // Sync product to SyncCreate for creative generation
    rpc SyncProductForCreatives(SyncProductRequest) returns (SyncProductResponse);
    
    // Batch sync for catalog-wide creative updates
    rpc BatchSyncProducts(BatchSyncProductsRequest) returns (BatchSyncProductsResponse);
    
    // Trigger creative generation when new product added
    rpc TriggerCreativeGeneration(TriggerCreativeRequest) returns (TriggerCreativeResponse);
}

message SyncProductRequest {
    int32 store_id = 1;
    Product product = 2;
    bool auto_generate_creatives = 3;
}

message SyncProductResponse {
    bool success = 1;
    int32 creative_assets_generated = 2;
    repeated string asset_urls = 3;
    string error = 4;
}

message BatchSyncProductsRequest {
    int32 store_id = 1;
    repeated Product products = 2;
}

message BatchSyncProductsResponse {
    bool success = 1;
    int32 products_synced = 2;
    int32 total_creatives_generated = 3;
    string error = 4;
}

message TriggerCreativeRequest {
    int32 store_id = 1;
    int32 product_id = 2;
    repeated string creative_types = 3; // "image", "video", "carousel"
    string brand_guidelines_id = 4;
}

message TriggerCreativeResponse {
    bool success = 1;
    string job_id = 2;
    int32 estimated_assets = 3;
    string error = 4;
}

// ============================================================================
// SyncEngage™ Integration (Retention Flows)
// ============================================================================

service RetentionFlowService {
    // Trigger post-purchase nurture flow
    rpc TriggerPostPurchaseFlow(TriggerFlowRequest) returns (TriggerFlowResponse);
    
    // Trigger churn prevention flow for high-risk customers
    rpc TriggerChurnPreventionFlow(TriggerFlowRequest) returns (TriggerFlowResponse);
    
    // Trigger upsell flow based on purchase history
    rpc TriggerUpsellFlow(TriggerFlowRequest) returns (TriggerFlowResponse);
}

message TriggerFlowRequest {
    int32 store_id = 1;
    int32 customer_id = 2;
    string flow_type = 3; // "post_purchase", "churn_prevention", "upsell"
    int32 order_id = 4;
    map<string, string> custom_fields = 5;
}

message TriggerFlowResponse {
    bool success = 1;
    string flow_execution_id = 2;
    int32 messages_scheduled = 3;
    string error = 4;
}

// ============================================================================
// SyncShield™ Integration (Token Encryption)
// ============================================================================

service TokenEncryptionService {
    // Encrypt OAuth tokens before storing in database
    rpc EncryptToken(EncryptTokenRequest) returns (EncryptTokenResponse);
    
    // Decrypt OAuth tokens for API calls
    rpc DecryptToken(DecryptTokenRequest) returns (DecryptTokenResponse);
    
    // Rotate encryption keys (monthly scheduled job)
    rpc RotateEncryptionKey(RotateKeyRequest) returns (RotateKeyResponse);
}

message EncryptTokenRequest {
    string plaintext_token = 1;
    string token_type = 2; // "access_token", "refresh_token", "api_key"
    int32 store_id = 3;
}

message EncryptTokenResponse {
    bool success = 1;
    string encrypted_token = 2;
    string encryption_key_id = 3;
    string error = 4;
}

message DecryptTokenRequest {
    string encrypted_token = 1;
    string encryption_key_id = 2;
    int32 store_id = 3;
}

message DecryptTokenResponse {
    bool success = 1;
    string plaintext_token = 2;
    string error = 3;
}

message RotateKeyRequest {
    string old_key_id = 1;
    string new_key_id = 2;
}

message RotateKeyResponse {
    bool success = 1;
    int32 tokens_re_encrypted = 2;
    string error = 3;
}

// ============================================================================
// Health Check & Status
// ============================================================================

service CMSIntegrationHealth {
    // Check if CMS connector is healthy
    rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
    
    // Get sync status for a store
    rpc GetSyncStatus(SyncStatusRequest) returns (SyncStatusResponse);
}

message HealthCheckRequest {
    string service_name = 1;
}

message HealthCheckResponse {
    bool healthy = 1;
    string version = 2;
    int64 uptime_seconds = 3;
    string error = 4;
}

message SyncStatusRequest {
    int32 store_id = 1;
}

message SyncStatusResponse {
    bool sync_enabled = 1;
    google.protobuf.Timestamp last_product_sync = 2;
    google.protobuf.Timestamp last_order_sync = 3;
    int32 pending_webhooks = 4;
    int32 failed_webhooks_24h = 5;
}
